> ### NOTEBOOK OBJECTIVE<br>
- Create a dataset of the sales records of the (not so) ficticious store `ROKO Shopping` for the year 2020.<br>

General steps:<br>
- Populate dataset with necessary features (columns) and data
- Clean, accomodate and put data toghether
- Save and export dataset

<hr>

#### Import Required Libraries

In [1]:
import pandas as pd
import numpy as np
import requests
import json
from datetime import datetime, timedelta
import random
import calendar
import re
import uuid
from faker import Faker

### Step 1. Populate dataset with necessary features

Sub-steps:
1) Define main DF features
2) Retrieve products from different files and create a DF for each
3) Concatenate the created DFs

- Define main DF features<br>
[ADD A SUB_CATEGORY COLUMN]

In [2]:
features = ["Purchase ID", "Purchase Datetime", 
            "Product ID", "Category", "Sub Category", 
            "Product Name", "Unit Price", "Quantity", 
            "Sales", "Payment Method", "Online", 
            "Shipping Address", "Shipping Method", "Product URL"]
df = pd.DataFrame(columns=features)

In [3]:
df

Empty DataFrame
Columns: [Purchase ID, Purchase Datetime, Product ID, Category, Sub Category, Product Name, Unit Price, Quantity, Sales, Payment Method, Online, Shipping Address, Shipping Method, Product URL]
Index: []

- Retrieve products data from different files and create a DF for each

Products come from these files:<br>
- /home/rohkoder29/Documents/year2022/python/data_science/mines/amzon_mobile_items_20190928.csv
- /home/rohkoder29/Documents/year2022/python/data_science/kaggle/marketing_sample_for_amazon_com-ecommerce__20200101_20200131__10k_data.csv
- /home/rohkoder29/Documents/year2022/python/data_science/cif_sales_analysis/data/data.xlsx

*File 1*

In [4]:
file1 = "/home/rohkoder29/Documents/year2022/python/data_science/mines/amzon_mobile_items_20190928.csv"
# load only relevant columns
cols = ["brand", "title", "prices", "url"]
df1 = pd.read_csv(file1, index_col=False, usecols=cols, skipinitialspace=True)

In [5]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 792 entries, 0 to 791
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   brand   792 non-null    object
 1   title   792 non-null    object
 2   url     792 non-null    object
 3   prices  577 non-null    object
dtypes: object(4)
memory usage: 24.9+ KB


In [6]:
df1.head()

brand                                              title  \
0     Nokia  Dual-Band / Tri-Mode Sprint PCS Phone w/ Voice...   
1  Motorola                                Motorola I265 phone   
2  Motorola  MOTOROLA C168i AT&T CINGULAR PREPAID GOPHONE C...   
3     Nokia  Nokia 6500 Slide Black/silver Unlocked Cell Phone   
4  Motorola              Motorola i335 Cell Phone Boost Mobile   

                                                 url  prices  
0  https://www.amazon.com/Dual-Band-Tri-Mode-Acti...     NaN  
1  https://www.amazon.com/Motorola-i265-I265-phon...  $49.95  
2  https://www.amazon.com/MOTOROLA-C168i-CINGULAR...     NaN  
3  https://www.amazon.com/Nokia-6500-Slide-silver...     NaN  
4  https://www.amazon.com/Motorola-i335-Phone-Boo...     NaN

Drop rows where price is missing

In [7]:
df1 = df1[df1["prices"].notna()]

In [8]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 577 entries, 1 to 791
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   brand   577 non-null    object
 1   title   577 non-null    object
 2   url     577 non-null    object
 3   prices  577 non-null    object
dtypes: object(4)
memory usage: 22.5+ KB


In [9]:
df1.tail()

brand                                              title  \
787     Apple  Apple iPhone XS, 512GB, Gold - For Verizon (Re...   
788   Samsung  Samsung Galaxy A50 128GB, 4GB RAM 6.4" Display...   
789   Samsung  Samsung Galaxy S9 (64GB, 4GB RAM) 5.8" QHD+ Di...   
790  Motorola  Motorola Moto G7 Play (32GB, 2GB RAM) 5.7" HD+...   
791    HUAWEI  Honor 5X Unlocked Smartphone, 16GB Dark Grey (...   

                                                   url             prices  
787  https://www.amazon.com/Apple-iPhone-XS-512GB-G...  $999.95,$1,349.99  
788  https://www.amazon.com/Samsung-A50-Display-Fac...            $304.99  
789  https://www.amazon.com/Samsung-Galaxy-S9-Displ...            $519.99  
790  https://www.amazon.com/Motorola-Vision-Display...            $139.99  
791  https://www.amazon.com/Honor-Unlocked-Smartpho...             $74.99

I've noticed that prices are stored as strings and some rows contain 2 prices. I'm gonna fix that.

In [10]:
def fix_two_prices(price):
    price = price.split(",")
    return max(price[0], price[1]) if len(price) > 1 else price[0]

In [11]:
df1["prices"] = df1["prices"].apply(lambda x: fix_two_prices(x))

Also change the type from Object to Float

In [12]:
df1["prices"] = df1["prices"].apply(lambda x: float(x[1:]))

In [13]:
# verify
df1["prices"].unique()

array([ 49.95,  78.99,  99.99,  59.89,  99.95, 149.95,  79.  ,  79.99,
        59.95,  59.99, 109.99, 118.8 , 189.95,  89.  , 389.28, 181.49,
       168.96, 164.92,  99.  ,  84.94, 230.59, 183.44, 165.59, 239.  ,
       209.65, 159.01, 142.99, 133.99, 169.  , 339.99, 494.99,  89.95,
       239.99, 219.99, 157.99, 214.5 , 159.99, 199.99,  28.69, 169.93,
       256.57, 249.99, 229.99, 339.47, 158.  , 199.  ,  80.  , 125.55,
       169.99, 153.  ,  79.95, 159.95, 149.99, 516.43, 159.  , 155.99,
       198.94, 104.  , 154.95, 119.  , 139.98, 151.99, 139.99, 129.99,
       130.  , 109.5 ,  78.74, 279.99, 110.95, 323.98, 112.95, 671.46,
       319.99, 138.  , 659.99, 145.  ,  95.  , 142.14, 299.99,  54.99,
       165.42, 138.99, 149.  ,  69.49,  94.99, 114.99, 259.98, 355.99,
       135.98, 139.89, 279.  , 140.55, 161.99, 219.53, 114.4 , 123.99,
       185.49, 135.55, 329.99, 179.99, 169.95, 269.99, 227.99, 217.99,
       145.99, 208.95,  88.95, 189.  , 240.  , 199.95, 144.  , 191.98,
      

In [14]:
df1.head(10)

brand                                              title  \
1   Motorola                                Motorola I265 phone   
6       Sony  Sony Ericsson G700 Triband GSM Phone Bronze (U...   
8      Nokia                        Nokia New 1100 for Tracfone   
9    Samsung             Samsung T301G Prepaid Phone (Tracfone)   
10  Motorola              Motorola I205 cell phone nextel/Boost   
11   Samsung                Samsung a167 Prepaid GoPhone (AT&T)   
16  Motorola  Motorola i570 Nextel iDen PTT rugged black cel...   
19   Samsung                     Samsung R355C Net 10 Unlimited   
24  Motorola  Motorola Clutch i475 Prepaid Phone (Boost Mobile)   
26    HUAWEI           Huawei Pinnacle Prepaid Phone (MetroPCS)   

                                                  url  prices  
1   https://www.amazon.com/Motorola-i265-I265-phon...   49.95  
6   https://www.amazon.com/Sony-Ericsson-Triband-B...   78.99  
8   https://www.amazon.com/New-Nokia-1100-for-Trac...   99.99  
9   https://www.amazon.com/Samsung-T301G-Prepaid-P...   59.89  
10  https://www.amazon.com/Motorola-I205-phone-nex...   99.95  
11  https://www.amazon.com/Samsung-a167-Prepaid-Go...  149.95  
16  https://www.amazon.com/Motorola-i570-Nextel-ru...   79.00  
19  https://www.amazon.com/Samsung-R355C-Net-10-Un...   79.99  
24  https://www.amazon.com/Motorola-Clutch-i475-Pr...   59.95  
26  https://www.amazon.com/Huawei-Pinnacle-Prepaid...   59.99

Consider only products within a certain price tag

In [15]:
df1 = df1.loc[df1["prices"] >= 300.00]

In [16]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 161 entries, 40 to 789
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   brand   161 non-null    object 
 1   title   161 non-null    object 
 2   url     161 non-null    object 
 3   prices  161 non-null    float64
dtypes: float64(1), object(3)
memory usage: 6.3+ KB


In [17]:
df1["title"].unique()[:50]

array(['Samsung Galaxy Express, Gray 8GB (AT&T)',
       'Samsung Galaxy Alpha, Charcoal Black 32GB (AT&T)',
       'Samsung Galaxy Mega 2, Brown Black 16GB (AT&T)',
       'Samsung Galaxy S6 Edge G925F 32GB Unlocked Phone - Retail Packaging - Black Sapphire',
       'Samsung Galaxy S6 Active, 32 GB , Grey (AT&T)',
       'Samsung Galaxy S7 Edge SM-G935F 32GB (GSM Only, No CDMA) Factory Unlocked 4G/LTE Single SIM Smartphone (Black Onyx)',
       'Samsung Galaxy GS7 Edge, Gold 32GB (Verizon Wireless)',
       'Samsung Galaxy S7 G930F 32GB Factory Unlocked GSM Smartphone - International Version - Titanium Silver',
       'Apple iPhone 6S, GSM Unlocked, 64GB - Rose Gold (Renewed)',
       'Samsung Galaxy S7 Edge unlocked smartphone, 32 GB Silver (US Warranty - Model SM-G935UZSAXAA)',
       'Samsung Galaxy S7 SM-G930F 32GB Factory Unlocked GSM 4G LTE Single Sim Smartphone (Black)',
       'Apple iPhone SE, 64GB, Gold - For AT&T / T-Mobile (Renewed)',
       'Google Pixel XL (128GB, 4GB RA

Maybe we should remove the (Renewed), (Refurbished)... from the items names

In [18]:
def clean_item_name(item: str) -> str:
    return item.split(" (")[0]

In [19]:
df1["title"] = df1["title"].apply(lambda x: clean_item_name(x))

Shorten all the URLs [using API]

In [20]:
endpoint = "https://api.shrtco.de/v2/shorten?url="
def shorten_url(url: str) -> str:
    response = requests.get(f"{endpoint}{url}").text
    if "result" in json.loads(response):
        return json.loads(response)["result"]["short_link2"]
    else:
        return "NA"

In [21]:
# df1["url"] = df1["url"].apply(lambda x: shorten_url(x))

In [22]:
df1.tail()

brand                                              title  \
781  Samsung            Samsung Galaxy A80 SM-A805F/DS Dual Sim   
786  Samsung  Samsung Galaxy A50 US Version Factory Unlocked...   
787    Apple         Apple iPhone XS, 512GB, Gold - For Verizon   
788  Samsung  Samsung Galaxy A50 128GB, 4GB RAM 6.4" Display...   
789  Samsung                                  Samsung Galaxy S9   

                                                   url  prices  
781  https://www.amazon.com/Samsung-A80-SM-A805F-Fa...  479.79  
786  https://www.amazon.com/Samsung-Unlocked-SM-A50...  349.99  
787  https://www.amazon.com/Apple-iPhone-XS-512GB-G...  999.95  
788  https://www.amazon.com/Samsung-A50-Display-Fac...  304.99  
789  https://www.amazon.com/Samsung-Galaxy-S9-Displ...  519.99

In [23]:
df1.reset_index(drop=True, inplace=True)

*File 2*

In [24]:
file2 = "/home/rohkoder29/Documents/year2022/python/data_science/kaggle/marketing_sample_for_amazon_com-ecommerce__20200101_20200131__10k_data.csv"
# load only relevant columns
cols = ["Product Name", "Category", "Selling Price", "Product Url"]
df2 = pd.read_csv(file2, index_col=False, skipinitialspace=True, usecols=cols)

In [25]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10002 entries, 0 to 10001
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Product Name   10002 non-null  object
 1   Category       9172 non-null   object
 2   Selling Price  9895 non-null   object
 3   Product Url    10002 non-null  object
dtypes: object(4)
memory usage: 312.7+ KB


Drop rows where Category or Price is missing

In [26]:
df2 = df2.loc[df2["Category"].notna()]
df2 = df2.loc[df2["Selling Price"].notna()]

In [27]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9066 entries, 0 to 10001
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Product Name   9066 non-null   object
 1   Category       9066 non-null   object
 2   Selling Price  9066 non-null   object
 3   Product Url    9066 non-null   object
dtypes: object(4)
memory usage: 354.1+ KB


In [28]:
df2.head()

Product Name  \
0  DB Longboards CoreFlex Crossbow 41" Bamboo Fib...   
1  Electronic Snap Circuits Mini Kits Classpack, ...   
2  3Doodler Create Flexy 3D Printing Filament Ref...   
3  Guillow Airplane Design Studio with Travel Cas...   
4                   Woodstock- Collage 500 pc Puzzle   

                                            Category Selling Price  \
0  Sports & Outdoors | Outdoor Recreation | Skate...       $237.68   
1  Toys & Games | Learning & Education | Science ...        $99.95   
2          Toys & Games | Arts & Crafts | Craft Kits        $34.99   
3  Toys & Games | Hobbies | Models & Model Kits |...        $28.91   
4            Toys & Games | Puzzles | Jigsaw Puzzles        $17.49   

                                         Product Url  
0  https://www.amazon.com/DB-Longboards-CoreFlex-...  
1  https://www.amazon.com/Electronic-Circuits-Cla...  
2  https://www.amazon.com/3Doodler-Plastic-Innova...  
3  https://www.amazon.com/Guillow-Airplane-Design...  
4  https://www.amazon.com/Woodstock-Collage-500-p...

Let's filter out...

We gonna keep products whitin only certain categories and sub-categories

Create a new Sub-Category feature

In [29]:
df2["Sub-Category"] = df2["Category"].apply(lambda x: x.split(" | ")[1])

Category feature has only the main categories

In [30]:
df2["Category"] = df2["Category"].apply(lambda x: x.split(" | ")[0])

In [31]:
df2.head()

Product Name           Category  \
0  DB Longboards CoreFlex Crossbow 41" Bamboo Fib...  Sports & Outdoors   
1  Electronic Snap Circuits Mini Kits Classpack, ...       Toys & Games   
2  3Doodler Create Flexy 3D Printing Filament Ref...       Toys & Games   
3  Guillow Airplane Design Studio with Travel Cas...       Toys & Games   
4                   Woodstock- Collage 500 pc Puzzle       Toys & Games   

  Selling Price                                        Product Url  \
0       $237.68  https://www.amazon.com/DB-Longboards-CoreFlex-...   
1        $99.95  https://www.amazon.com/Electronic-Circuits-Cla...   
2        $34.99  https://www.amazon.com/3Doodler-Plastic-Innova...   
3        $28.91  https://www.amazon.com/Guillow-Airplane-Design...   
4        $17.49  https://www.amazon.com/Woodstock-Collage-500-p...   

           Sub-Category  
0    Outdoor Recreation  
1  Learning & Education  
2         Arts & Crafts  
3               Hobbies  
4               Puzzles

In [32]:
df2["Category"].unique()

array(['Sports & Outdoors', 'Toys & Games', 'Clothing, Shoes & Jewelry',
       'Home & Kitchen', 'Baby Products', 'Pet Supplies',
       'Arts, Crafts & Sewing', 'Electronics', 'Video Games',
       'Patio, Lawn & Garden', 'Tools & Home Improvement',
       'Office Products', 'Grocery & Gourmet Food',
       'Remote & App Controlled Vehicle Parts', 'Health & Household',
       'Industrial & Scientific', 'Beauty & Personal Care', 'Hobbies',
       'Remote & App Controlled Vehicles & Parts', 'Automotive',
       'Musical Instruments', 'Movies & TV', 'Cell Phones & Accessories'],
      dtype=object)

In [33]:
df2["Sub-Category"].unique()

array(['Outdoor Recreation', 'Learning & Education', 'Arts & Crafts',
       'Hobbies', 'Puzzles', 'Costumes & Accessories', 'Home Décor',
       'Baby & Toddler Toys', 'Collectible Toys', 'Nursery',
       'Building Toys', 'Furniture', 'Novelty & Gag Toys',
       'Sports & Outdoor Play', 'Party Supplies',
       'Tricycles, Scooters & Wagons', "Kids' Electronics",
       'Sports & Fitness', 'Dress Up & Pretend Play', 'Bedding',
       'Toy Figures & Playsets', 'Play Vehicles', 'Dolls & Accessories',
       'Games & Accessories', 'Dogs', 'Novelty & More', 'Men',
       'Stuffed Animals & Plush Toys', 'Fan Shop', 'Crafting',
       'Camera & Photo', 'Beading & Jewelry Making', 'PlayStation 4',
       'Baby Care', 'Outdoor Décor', 'Power & Hand Tools',
       'Painting, Drawing & Art Supplies', 'Lighting & Ceiling Fans',
       'Office & School Supplies', 'Activity & Entertainment',
       'Nintendo Switch', 'Luggage & Travel Gear',
       "Kids' Furniture, Décor & Storage", 'Strollers 

In [34]:
df2 = df2.loc[(df2["Category"].str.contains("electronics|office|accessories|^video games",
                                     regex=True, flags=re.I)) |
        (df2["Sub-Category"].str.contains("learning|education|^accessories|games|tools|^kids",
                                     regex=True, flags=re.I))]

In [35]:
df2.head()

Product Name      Category  \
1   Electronic Snap Circuits Mini Kits Classpack, ...  Toys & Games   
16  The Complete Common Core: State Standards Kit,...  Toys & Games   
26  NKOK Sonic and Sega All Stars Racing Remote Co...  Toys & Games   
42  Random Esoteric Creature Generator for Classic...  Toys & Games   
48  Melissa & Doug Disney Cinderella Rags to Riche...  Toys & Games   

   Selling Price                                        Product Url  \
1         $99.95  https://www.amazon.com/Electronic-Circuits-Cla...   
16        $12.88  https://www.amazon.com/Complete-Common-Core-St...   
26        $26.99  https://www.amazon.com/NKOK-Sonic-Racing-Remot...   
42        $19.94  https://www.amazon.com/Esoteric-Creature-Gener...   
48        $11.12  https://www.amazon.com/Melissa-Doug-Disney-Cin...   

            Sub-Category  
1   Learning & Education  
16  Learning & Education  
26     Kids' Electronics  
42   Games & Accessories  
48  Learning & Education

In [36]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1269 entries, 1 to 9999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Product Name   1269 non-null   object
 1   Category       1269 non-null   object
 2   Selling Price  1269 non-null   object
 3   Product Url    1269 non-null   object
 4   Sub-Category   1269 non-null   object
dtypes: object(5)
memory usage: 59.5+ KB


+ Convert Price to float

In [37]:
df2["Selling Price"].unique()

array(['$99.95', '$12.88', '$26.99', '$19.94', '$11.12', '$7.99', '$6.11',
       '$38.33', '$16.43', '$14.01', '$10.99', '$186.06', '$8.53',
       '$38.55', '$6.47', '$107.86', '$12.50', '$36.91', '$11.24',
       '$10.94',
       '$ 19 99 $39.95 #listPriceLegalMessageText { margin-left: 4px !important; } #listPriceLegalMessage .a-popover-trigger:hover { text-decoration: none !important; } #listPriceLegalMessage .a-icon-popover { display: none !important; margin-left: 0px !important; margin-top: 6px !important; } Save $19.96 (50%)',
       '$4.99', '$9.99', '$14.99', '$6.96', '$9.95', '$12.47', '$8.29',
       '$17.40', '$6.63', '$9.96', '$6.69', '$12.64', '$25.39', '$32.84',
       '$35.00', '$8.91', '$143.99', '$10.44 $ 10 . 44', '$52.44',
       '$19.99', '$3.49', '$20.18', '$89.95', '$89.99', '$24.99',
       '$16.78', '$39.99', '$15.51', '$10.62', '$15.20', '$81.54',
       '$12.91', '$17.63', '$14.90', '$34.95', '$65.89', '$43.00',
       '$3.07', '$24.49', '$18.99', '$17.49', 

Seems like we'll have to clean up the prices before we can convert them to float

Set a pattern for wich prices are valid

In [38]:
df2 = df2.loc[df2["Selling Price"].str.contains("^\$[0-9]+?.[0-9]+?$",
                                          regex=True, flags=re.I)]

In [39]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1239 entries, 1 to 9999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Product Name   1239 non-null   object
 1   Category       1239 non-null   object
 2   Selling Price  1239 non-null   object
 3   Product Url    1239 non-null   object
 4   Sub-Category   1239 non-null   object
dtypes: object(5)
memory usage: 58.1+ KB


In [40]:
# finally convert to float
df2["Selling Price"] = df2["Selling Price"].apply(lambda x: float(x.split()[0][1:]))

In [41]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1239 entries, 1 to 9999
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Product Name   1239 non-null   object 
 1   Category       1239 non-null   object 
 2   Selling Price  1239 non-null   float64
 3   Product Url    1239 non-null   object 
 4   Sub-Category   1239 non-null   object 
dtypes: float64(1), object(4)
memory usage: 58.1+ KB


Last but not least, let's shorten the URLs

In [42]:
# using the same old function
# df2["Product Url"] = df2["Product Url"].apply(lambda x: shorten_url(x))

In [43]:
df2.reset_index(drop=True, inplace=True)

In [44]:
df2.tail()

Product Name  \
1234                      Green Couch Games Filler Game   
1235                                Dominion: Menagerie   
1236  Sax True Flow Spectrum Watercolor Paint Brushe...   
1237  Remedia Publications REM536B Money Activity Bo...   
1238  NewPath Learning 10 Piece Science Owls and Owl...   

                     Category  Selling Price  \
1234             Toys & Games          15.00   
1235             Toys & Games          44.95   
1236  Industrial & Scientific          19.05   
1237             Toys & Games           9.31   
1238          Office Products          37.95   

                                            Product Url  \
1234  https://www.amazon.com/Green-Couch-Games-Fille...   
1235  https://www.amazon.com/Rio-Grande-Games-Domini...   
1236  https://www.amazon.com/Sax-Spectrum-Watercolor...   
1237  https://www.amazon.com/Remedia-Publications-RE...   
1238  https://www.amazon.com/NewPath-Learning-Scienc...   

                  Sub-Category  
1234       Games & Accessories  
1235       Games & Accessories  
1236         Science Education  
1237      Learning & Education  
1238  Office & School Supplies

*File 3*

In [45]:
file3 = "/home/rohkoder29/Documents/year2022/python/data_science/cif_sales_analysis/data/data.xlsx"
cols = ["Category", "Sub-Category", "Product Name", "Sales", "Quantity"]
df3 = pd.read_excel(file3, index_col=False, usecols=cols)

In [46]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3312 entries, 0 to 3311
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Category      3312 non-null   object 
 1   Sub-Category  3312 non-null   object 
 2   Product Name  3312 non-null   object 
 3   Sales         3312 non-null   float64
 4   Quantity      3312 non-null   int64  
dtypes: float64(1), int64(1), object(3)
memory usage: 129.5+ KB


Cool, no missing values.

In [47]:
df3.head()

Category Sub-Category  \
0  Office Supplies        Paper   
1        Furniture       Chairs   
2  Office Supplies        Paper   
3       Technology       Phones   
4  Office Supplies      Storage   

                                        Product Name    Sales  Quantity  
0                                         Xerox 1967   15.552         3  
1                 Global Deluxe Stacking Chair, Gray   71.372         2  
2                                  Easy-staple paper   29.472         3  
3                                 Panasonic Kx-TS550  147.168         4  
4  Advantus 10-Drawer Portable Organizer, Chrome ...   95.616         2

In [48]:
df3["Category"].unique()

array(['Office Supplies', 'Furniture', 'Technology'], dtype=object)

In [49]:
df3["Sub-Category"].unique()

array(['Paper', 'Chairs', 'Phones', 'Storage', 'Binders', 'Furnishings',
       'Accessories', 'Labels', 'Art', 'Fasteners', 'Envelopes',
       'Appliances', 'Tables', 'Copiers', 'Bookcases', 'Supplies',
       'Machines'], dtype=object)

We don't need the Furniture category, so we drop it

In [50]:
df3 = df3.loc[df3["Category"].isin(["Office Supplies", "Technology"])]

In [51]:
df3.tail()

Category Sub-Category  \
3306       Technology  Accessories   
3307       Technology       Phones   
3309       Technology       Phones   
3310  Office Supplies        Paper   
3311  Office Supplies   Appliances   

                                           Product Name    Sales  Quantity  
3306  Logitech G430 Surround Sound Gaming Headset wi...   79.990         1  
3307                    Panasonic KX - TS880B Telephone  206.100         5  
3309                              Aastra 57i VoIP phone  258.576         2  
3310  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.600         4  
3311  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.160         2

In fact we could get the price for each product, given we have sales and quantity

In [52]:
df3["Unit_Price"] = df3["Sales"] / df3["Quantity"]

In [53]:
df3.tail()

Category Sub-Category  \
3306       Technology  Accessories   
3307       Technology       Phones   
3309       Technology       Phones   
3310  Office Supplies        Paper   
3311  Office Supplies   Appliances   

                                           Product Name    Sales  Quantity  \
3306  Logitech G430 Surround Sound Gaming Headset wi...   79.990         1   
3307                    Panasonic KX - TS880B Telephone  206.100         5   
3309                              Aastra 57i VoIP phone  258.576         2   
3310  It's Hot Message Books with Stickers, 2 3/4" x 5"   29.600         4   
3311  Acco 7-Outlet Masterpiece Power Center, Wihtou...  243.160         2   

      Unit_Price  
3306      79.990  
3307      41.220  
3309     129.288  
3310       7.400  
3311     121.580

Now that we've got our DFs it's time for some cleaning.

### Step 3. Clean & Accomodate the DFs

Sub-steps:<br>
- [ADD A SUB_CATEGORY FEATURE IN ALL DFs]
- In DF1, replace the `brand` column by a `Sub_Category` one [with value = "Phones" so it matches the existing one in other DFs] and rename the `title` column to `Product_Name` and add `Quantity` and `Sales` column that we'll fill in later on
- DF2 is almost okay, again we need to take care of the `Quantity` add `Sales` features
- In DF3 we rearrange the columns order and add a `Product_URL` column at the end

*DF1*

In [54]:
df1.head(1)

brand                             title  \
0  Samsung  Samsung Galaxy Express, Gray 8GB   

                                                 url  prices  
0  https://www.amazon.com/Samsung-Galaxy-Express-...  389.28

In [55]:
df1["brand"] = df1["brand"].apply(lambda x: x.replace(x, "Phones"))

Rename and reindex the columns to match main DF

In [56]:
df1.rename(columns={"brand": "Sub_Category",
                    "title": "Product_Name",
                    "prices": "Unit_Price",
                    "url": "Product_URL"}, inplace=True)

In [57]:
df1 = df1.reindex(columns=["Category", "Sub_Category", "Product_Name", "Unit_Price", "Quantity", "Sales", "Product_URL"])

In [58]:
df1["Category"] = df1["Category"].astype(np.str0)

In [59]:
df1["Category"] = df1["Category"].apply(lambda x: x.replace(x, "Electronics"))

In [60]:
df1.head(1)

Category Sub_Category                      Product_Name  Unit_Price  \
0  Electronics       Phones  Samsung Galaxy Express, Gray 8GB      389.28   

   Quantity  Sales                                        Product_URL  
0       NaN    NaN  https://www.amazon.com/Samsung-Galaxy-Express-...

*DF2*

In [61]:
df2.head(1)

Product Name      Category  \
0  Electronic Snap Circuits Mini Kits Classpack, ...  Toys & Games   

   Selling Price                                        Product Url  \
0          99.95  https://www.amazon.com/Electronic-Circuits-Cla...   

           Sub-Category  
0  Learning & Education

Rename and reindex the columns to match main DF

In [62]:
df2.rename(columns={"Product Name": "Product_Name",
                    "Selling Price": "Unit_Price",
                    "Product Url": "Product_URL",
                    "Sub-Category": "Sub_Category"}, inplace=True)

In [63]:
df2 = df2.reindex(columns=["Category", "Sub_Category", "Product_Name", "Unit_Price", "Quantity", "Sales", "Product_URL"])

In [64]:
df2.head(1)

Category          Sub_Category  \
0  Toys & Games  Learning & Education   

                                        Product_Name  Unit_Price  Quantity  \
0  Electronic Snap Circuits Mini Kits Classpack, ...       99.95       NaN   

   Sales                                        Product_URL  
0    NaN  https://www.amazon.com/Electronic-Circuits-Cla...

*DF3*

In [65]:
df3.head()

Category Sub-Category  \
0  Office Supplies        Paper   
2  Office Supplies        Paper   
3       Technology       Phones   
4  Office Supplies      Storage   
5  Office Supplies        Paper   

                                        Product Name    Sales  Quantity  \
0                                         Xerox 1967   15.552         3   
2                                  Easy-staple paper   29.472         3   
3                                 Panasonic Kx-TS550  147.168         4   
4  Advantus 10-Drawer Portable Organizer, Chrome ...   95.616         2   
5  Telephone Message Books with Fax/Mobile Sectio...   19.050         3   

   Unit_Price  
0       5.184  
2       9.824  
3      36.792  
4      47.808  
5       6.350

In [66]:
# let's change Technology to Electronics
df3["Category"] = df3["Category"].apply(lambda x: x.replace("Technology", "Electronics"))

Rename and reindex the columns to match main DF

NOTE: I think I'm gonna keep the Sub-Category feature, and accomodate the other DFs later on.

In [67]:
# rename Product Name
df3.rename(columns={"Product Name": "Product_Name",
                    "Sub-Category": "Sub_Category"}, inplace=True)

In [68]:
df3 = df3.reindex(columns=["Category", "Sub_Category", "Product_Name", "Unit_Price", "Quantity", "Sales", "Product_URL"])

In [69]:
df3.head()

Category Sub_Category  \
0  Office Supplies        Paper   
2  Office Supplies        Paper   
3      Electronics       Phones   
4  Office Supplies      Storage   
5  Office Supplies        Paper   

                                        Product_Name  Unit_Price  Quantity  \
0                                         Xerox 1967       5.184         3   
2                                  Easy-staple paper       9.824         3   
3                                 Panasonic Kx-TS550      36.792         4   
4  Advantus 10-Drawer Portable Organizer, Chrome ...      47.808         2   
5  Telephone Message Books with Fax/Mobile Sectio...       6.350         3   

     Sales  Product_URL  
0   15.552          NaN  
2   29.472          NaN  
3  147.168          NaN  
4   95.616          NaN  
5   19.050          NaN

Sweet!

In [70]:
df1.tail(1)

Category Sub_Category       Product_Name  Unit_Price  Quantity  Sales  \
160  Electronics       Phones  Samsung Galaxy S9      519.99       NaN    NaN   

                                           Product_URL  
160  https://www.amazon.com/Samsung-Galaxy-S9-Displ...

In [71]:
df2.tail(1)

Category              Sub_Category  \
1238  Office Products  Office & School Supplies   

                                           Product_Name  Unit_Price  Quantity  \
1238  NewPath Learning 10 Piece Science Owls and Owl...       37.95       NaN   

      Sales                                        Product_URL  
1238    NaN  https://www.amazon.com/NewPath-Learning-Scienc...

In [72]:
df3.tail(1)

Category Sub_Category  \
3311  Office Supplies   Appliances   

                                           Product_Name  Unit_Price  Quantity  \
3311  Acco 7-Outlet Masterpiece Power Center, Wihtou...      121.58         2   

       Sales  Product_URL  
3311  243.16          NaN

Quick notice: in DF2 the category is "Office Products" whereas in DF3 it is "Office Supplies", let's change the former to the latter

In [73]:
df2["Category"] = df2["Category"].apply(lambda x: x.replace("Office Products", "Office Supplies"))

- So far so good. Now let's combine them!

In [74]:
df4 = pd.concat([df1, df2, df3], ignore_index=True).sort_values(by=["Category"]).reset_index(drop=True)

In [75]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Category      4026 non-null   object 
 1   Sub_Category  4026 non-null   object 
 2   Product_Name  4026 non-null   object 
 3   Unit_Price    4026 non-null   float64
 4   Quantity      2626 non-null   float64
 5   Sales         2626 non-null   float64
 6   Product_URL   1400 non-null   object 
dtypes: float64(3), object(4)
memory usage: 220.3+ KB


*How cool is that? [in Craig Fedherighi voice]*

In [76]:
df4.head()

Category         Sub_Category  \
0     Beauty & Personal Care  Tools & Accessories   
1  Cell Phones & Accessories          Accessories   
2                Electronics               Phones   
3                Electronics               Phones   
4                Electronics          Accessories   

                                        Product_Name  Unit_Price  Quantity  \
0                Rubies Black Eyelashes and Adhesive        7.15       NaN   
1  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...        8.25       NaN   
2                   Samsung Galaxy Express, Gray 8GB      389.28       NaN   
3                                     Apple iPhone 5      649.83       2.0   
4               SanDisk Cruzer 64 GB USB Flash Drive       36.32       2.0   

     Sales                                        Product_URL  
0      NaN  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1      NaN  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2      NaN  https://www.amazon.com/Samsung-Galaxy-Express-...  
3  1299.66                                                NaN  
4    72.64                                                NaN

- Replace all missing values in `Unit_Price`, `Quantity` (convert to int) and `Sales` by 0

In [77]:
df4["Unit_Price"].fillna(0, inplace=True)
df4["Quantity"] = df4["Quantity"].fillna(0).astype(np.int8)
df4["Sales"].fillna(0, inplace=True)

In [78]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Category      4026 non-null   object 
 1   Sub_Category  4026 non-null   object 
 2   Product_Name  4026 non-null   object 
 3   Unit_Price    4026 non-null   float64
 4   Quantity      4026 non-null   int8   
 5   Sales         4026 non-null   float64
 6   Product_URL   1400 non-null   object 
dtypes: float64(2), int8(1), object(4)
memory usage: 192.8+ KB


In [79]:
df4.head()

Category         Sub_Category  \
0     Beauty & Personal Care  Tools & Accessories   
1  Cell Phones & Accessories          Accessories   
2                Electronics               Phones   
3                Electronics               Phones   
4                Electronics          Accessories   

                                        Product_Name  Unit_Price  Quantity  \
0                Rubies Black Eyelashes and Adhesive        7.15         0   
1  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...        8.25         0   
2                   Samsung Galaxy Express, Gray 8GB      389.28         0   
3                                     Apple iPhone 5      649.83         2   
4               SanDisk Cruzer 64 GB USB Flash Drive       36.32         2   

     Sales                                        Product_URL  
0     0.00  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1     0.00  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2     0.00  https://www.amazon.com/Samsung-Galaxy-Express-...  
3  1299.66                                                NaN  
4    72.64                                                NaN

In [80]:
# just to be safe
# df4.to_csv("df4", index=False)

Now we have to combine DF4 with DF

But first let's rename the columns

In [81]:
old_cols = df.columns.tolist()
new_cols = [column.replace(" ", "_") for column in old_cols]

In [82]:
df.rename(columns=dict(zip(old_cols, new_cols)), inplace=True)

In [83]:
df

Empty DataFrame
Columns: [Purchase_ID, Purchase_Datetime, Product_ID, Category, Sub_Category, Product_Name, Unit_Price, Quantity, Sales, Payment_Method, Online, Shipping_Address, Shipping_Method, Product_URL]
Index: []

In [84]:
df5 = pd.concat([df, df4], ignore_index=True)

#### Now let's work with DF5

In [85]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Purchase_ID        0 non-null      object
 1   Purchase_Datetime  0 non-null      object
 2   Product_ID         0 non-null      object
 3   Category           4026 non-null   object
 4   Sub_Category       4026 non-null   object
 5   Product_Name       4026 non-null   object
 6   Unit_Price         4026 non-null   object
 7   Quantity           4026 non-null   object
 8   Sales              4026 non-null   object
 9   Payment_Method     0 non-null      object
 10  Online             0 non-null      object
 11  Shipping_Address   0 non-null      object
 12  Shipping_Method    0 non-null      object
 13  Product_URL        1400 non-null   object
dtypes: object(14)
memory usage: 440.5+ KB


In [86]:
df5.head(1)

Purchase_ID Purchase_Datetime Product_ID                Category  \
0         NaN               NaN        NaN  Beauty & Personal Care   

          Sub_Category                         Product_Name Unit_Price  \
0  Tools & Accessories  Rubies Black Eyelashes and Adhesive       7.15   

  Quantity Sales Payment_Method Online Shipping_Address Shipping_Method  \
0        0   0.0            NaN    NaN              NaN             NaN   

                                         Product_URL  
0  https://www.amazon.com/Rubies-Black-Eyelashes-...

- Generating data for columns `Purchase_ID`, `Purchase_Datetime`, `Online`, `Payment_Method`, `Shipping_Address`, `Shipping_Method`

*Online*<br>
Consideration: People usually buy goods within a certain price tag online, so we gonna consider only purchases over $69.99 to be elligible.

In order to perform this operation, we need to convert Sales column to float

In [87]:
df5["Sales"] = df5["Sales"].astype(np.float64)

In [88]:
# and keep only two decimals
df5["Sales"] = df5["Sales"].apply(lambda x: float(f"{x:.2f}"))

In [89]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Purchase_ID        0 non-null      object 
 1   Purchase_Datetime  0 non-null      object 
 2   Product_ID         0 non-null      object 
 3   Category           4026 non-null   object 
 4   Sub_Category       4026 non-null   object 
 5   Product_Name       4026 non-null   object 
 6   Unit_Price         4026 non-null   object 
 7   Quantity           4026 non-null   object 
 8   Sales              4026 non-null   float64
 9   Payment_Method     0 non-null      object 
 10  Online             0 non-null      object 
 11  Shipping_Address   0 non-null      object 
 12  Shipping_Method    0 non-null      object 
 13  Product_URL        1400 non-null   object 
dtypes: float64(1), object(13)
memory usage: 440.5+ KB


In [90]:
def set_online(row) -> bool:
    return False if row["Sales"] < 69.99 else random.choices([True, False], weights=[.9, .1])[0]

In [91]:
df5["Online"] = df5.apply(lambda x: set_online(x), axis=1)

In [92]:
df5.head()

Purchase_ID Purchase_Datetime Product_ID                   Category  \
0         NaN               NaN        NaN     Beauty & Personal Care   
1         NaN               NaN        NaN  Cell Phones & Accessories   
2         NaN               NaN        NaN                Electronics   
3         NaN               NaN        NaN                Electronics   
4         NaN               NaN        NaN                Electronics   

          Sub_Category                                       Product_Name  \
0  Tools & Accessories                Rubies Black Eyelashes and Adhesive   
1          Accessories  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...   
2               Phones                   Samsung Galaxy Express, Gray 8GB   
3               Phones                                     Apple iPhone 5   
4          Accessories               SanDisk Cruzer 64 GB USB Flash Drive   

  Unit_Price Quantity    Sales Payment_Method  Online Shipping_Address  \
0       7.15        0     0.00            NaN   False              NaN   
1       8.25        0     0.00            NaN   False              NaN   
2     389.28        0     0.00            NaN   False              NaN   
3     649.83        2  1299.66            NaN    True              NaN   
4      36.32        2    72.64            NaN    True              NaN   

  Shipping_Method                                        Product_URL  
0             NaN  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1             NaN  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2             NaN  https://www.amazon.com/Samsung-Galaxy-Express-...  
3             NaN                                                NaN  
4             NaN                                                NaN

In [93]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Purchase_ID        0 non-null      object 
 1   Purchase_Datetime  0 non-null      object 
 2   Product_ID         0 non-null      object 
 3   Category           4026 non-null   object 
 4   Sub_Category       4026 non-null   object 
 5   Product_Name       4026 non-null   object 
 6   Unit_Price         4026 non-null   object 
 7   Quantity           4026 non-null   object 
 8   Sales              4026 non-null   float64
 9   Payment_Method     0 non-null      object 
 10  Online             4026 non-null   bool   
 11  Shipping_Address   0 non-null      object 
 12  Shipping_Method    0 non-null      object 
 13  Product_URL        1400 non-null   object 
dtypes: bool(1), float64(1), object(12)
memory usage: 412.9+ KB


*Product_ID*<br>
Consideration: should contain characters indicating the item's category and sub-category

In [94]:
df5["Category"].unique()

array(['Beauty & Personal Care', 'Cell Phones & Accessories',
       'Electronics', 'Home & Kitchen', 'Industrial & Scientific',
       'Office Supplies', 'Tools & Home Improvement', 'Toys & Games',
       'Video Games'], dtype=object)

In [95]:
def get_cat_ini(category: str) -> str:
    ini = []
    ini.extend(part[0].upper() for part in category.split() if part[0].isalpha())
    return "".join(ini)

In [96]:
df5["Sub_Category"].unique()

array(['Tools & Accessories', 'Accessories', 'Phones', 'Machines',
       'Camera & Photo', 'Copiers', 'Headphones',
       'Portable Audio & Video', 'Computers & Accessories',
       "Kids' Home Store", 'Science Education', 'Binders', 'Paper',
       'Appliances', 'Art', 'Labels', 'Storage', 'Envelopes', 'Fasteners',
       'Supplies', 'Office & School Supplies', 'Office Electronics',
       'Power & Hand Tools', 'Games & Accessories', "Kids' Electronics",
       'Learning & Education', "Kids' Furniture, Décor & Storage",
       'Xbox One', 'Nintendo Switch', 'PC', 'Legacy Systems'],
      dtype=object)

*Well, while we're at it, I think products like Xbox One and Nintendo Switch should be placed under a `Console & Accessories` sub-category*

In [97]:
df5.loc[df5["Sub_Category"].str.contains("xbox|nintendo", case=False, regex=True), ["Sub_Category"]] = "Console & Accessories"

In [98]:
df5["Sub_Category"].unique()

array(['Tools & Accessories', 'Accessories', 'Phones', 'Machines',
       'Camera & Photo', 'Copiers', 'Headphones',
       'Portable Audio & Video', 'Computers & Accessories',
       "Kids' Home Store", 'Science Education', 'Binders', 'Paper',
       'Appliances', 'Art', 'Labels', 'Storage', 'Envelopes', 'Fasteners',
       'Supplies', 'Office & School Supplies', 'Office Electronics',
       'Power & Hand Tools', 'Games & Accessories', "Kids' Electronics",
       'Learning & Education', "Kids' Furniture, Décor & Storage",
       'Console & Accessories', 'PC', 'Legacy Systems'], dtype=object)

In [99]:
# este es una locura
df5["Product_ID"] = df5["Category"].apply(lambda x: get_cat_ini(x)) +"-"+ df5["Sub_Category"].apply(lambda x: get_cat_ini(x))

In [100]:
def randomize() -> str:
    return str(uuid.uuid4()).split("-")[-1].upper()

In [101]:
df5["Product_ID"] = df5["Product_ID"].apply(lambda x: f"{x}-{randomize()}")

In [102]:
df5.head()

Purchase_ID Purchase_Datetime           Product_ID  \
0         NaN               NaN  BPC-TA-595E364B36B8   
1         NaN               NaN   CPA-A-AD2CDBF8FB6A   
2         NaN               NaN     E-P-7B472F9085C3   
3         NaN               NaN     E-P-06BB0F1A9C0B   
4         NaN               NaN     E-A-2252DB4688C6   

                    Category         Sub_Category  \
0     Beauty & Personal Care  Tools & Accessories   
1  Cell Phones & Accessories          Accessories   
2                Electronics               Phones   
3                Electronics               Phones   
4                Electronics          Accessories   

                                        Product_Name Unit_Price Quantity  \
0                Rubies Black Eyelashes and Adhesive       7.15        0   
1  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...       8.25        0   
2                   Samsung Galaxy Express, Gray 8GB     389.28        0   
3                                     Apple iPhone 5     649.83        2   
4               SanDisk Cruzer 64 GB USB Flash Drive      36.32        2   

     Sales Payment_Method  Online Shipping_Address Shipping_Method  \
0     0.00            NaN   False              NaN             NaN   
1     0.00            NaN   False              NaN             NaN   
2     0.00            NaN   False              NaN             NaN   
3  1299.66            NaN    True              NaN             NaN   
4    72.64            NaN    True              NaN             NaN   

                                         Product_URL  
0  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2  https://www.amazon.com/Samsung-Galaxy-Express-...  
3                                                NaN  
4                                                NaN

*Purchase_Datetime*<br>
From Jan 1, 2020 to Dec 31, 2020 and 00:00 to 23:59 if online, or 00:09 to 21:59 if not

In [103]:
def gen_datetime(row) -> datetime:
    date_start = datetime(2020, 1, 1, 0, 0, 0)
    date_end = datetime(2020, 12, 31, 23, 59, 59)
    if row["Online"] is False:
        date_start = datetime(2020, 1, 1, 8, 0, 0)
        date_end = datetime(2020, 12, 31, 21, 59, 59)
    random_date_time = date_start + (date_end - date_start) * random.random()
    random_date_time = random_date_time.__format__("%Y-%m-%d %H:%M:%S")
    return datetime.strptime(random_date_time, format("%Y-%m-%d %H:%M:%S"))

In [104]:
df5["Purchase_Datetime"] = df5.apply(lambda x: gen_datetime(x), axis=1)

In [105]:
df5.head()

Purchase_ID   Purchase_Datetime           Product_ID  \
0         NaN 2020-11-13 23:00:27  BPC-TA-595E364B36B8   
1         NaN 2020-07-14 11:24:38   CPA-A-AD2CDBF8FB6A   
2         NaN 2020-03-18 13:04:50     E-P-7B472F9085C3   
3         NaN 2020-03-05 22:57:56     E-P-06BB0F1A9C0B   
4         NaN 2020-07-03 15:20:43     E-A-2252DB4688C6   

                    Category         Sub_Category  \
0     Beauty & Personal Care  Tools & Accessories   
1  Cell Phones & Accessories          Accessories   
2                Electronics               Phones   
3                Electronics               Phones   
4                Electronics          Accessories   

                                        Product_Name Unit_Price Quantity  \
0                Rubies Black Eyelashes and Adhesive       7.15        0   
1  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...       8.25        0   
2                   Samsung Galaxy Express, Gray 8GB     389.28        0   
3                                     Apple iPhone 5     649.83        2   
4               SanDisk Cruzer 64 GB USB Flash Drive      36.32        2   

     Sales Payment_Method  Online Shipping_Address Shipping_Method  \
0     0.00            NaN   False              NaN             NaN   
1     0.00            NaN   False              NaN             NaN   
2     0.00            NaN   False              NaN             NaN   
3  1299.66            NaN    True              NaN             NaN   
4    72.64            NaN    True              NaN             NaN   

                                         Product_URL  
0  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2  https://www.amazon.com/Samsung-Galaxy-Express-...  
3                                                NaN  
4                                                NaN

In [106]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Purchase_ID        0 non-null      object        
 1   Purchase_Datetime  4026 non-null   datetime64[ns]
 2   Product_ID         4026 non-null   object        
 3   Category           4026 non-null   object        
 4   Sub_Category       4026 non-null   object        
 5   Product_Name       4026 non-null   object        
 6   Unit_Price         4026 non-null   object        
 7   Quantity           4026 non-null   object        
 8   Sales              4026 non-null   float64       
 9   Payment_Method     0 non-null      object        
 10  Online             4026 non-null   bool          
 11  Shipping_Address   0 non-null      object        
 12  Shipping_Method    0 non-null      object        
 13  Product_URL        1400 non-null   object        
dtypes: bool(

In [107]:
df5.to_pickle("df5")

*Purchase_ID*<br>
Depends on `Purchase_Datetime` and `Online`<br>
Format: [100-999][datetime][online][100-999]

In [108]:
# concatenate Datetime
def conca_datetime(row):
    conca = []
    conca.extend(i for i in row if i.isdigit())
    return "".join(conca)

In [109]:
# add Online id
def add_online(row):
    return "O" if row["Online"] else "P"

In [110]:
def set_purchase_id(row) -> str:
    return f"{random.randint(100, 999)}{add_online(row)}{conca_datetime(str(row['Purchase_Datetime']))}{random.randint(100, 999)}"

In [111]:
df5["Purchase_ID"] = df5.apply(lambda x: set_purchase_id(x), axis=1)

In [112]:
df5.head()

Purchase_ID   Purchase_Datetime           Product_ID  \
0  116P20201113230027405 2020-11-13 23:00:27  BPC-TA-595E364B36B8   
1  412P20200714112438561 2020-07-14 11:24:38   CPA-A-AD2CDBF8FB6A   
2  284P20200318130450913 2020-03-18 13:04:50     E-P-7B472F9085C3   
3  210O20200305225756148 2020-03-05 22:57:56     E-P-06BB0F1A9C0B   
4  514O20200703152043999 2020-07-03 15:20:43     E-A-2252DB4688C6   

                    Category         Sub_Category  \
0     Beauty & Personal Care  Tools & Accessories   
1  Cell Phones & Accessories          Accessories   
2                Electronics               Phones   
3                Electronics               Phones   
4                Electronics          Accessories   

                                        Product_Name Unit_Price Quantity  \
0                Rubies Black Eyelashes and Adhesive       7.15        0   
1  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...       8.25        0   
2                   Samsung Galaxy Express, Gray 8GB     389.28        0   
3                                     Apple iPhone 5     649.83        2   
4               SanDisk Cruzer 64 GB USB Flash Drive      36.32        2   

     Sales Payment_Method  Online Shipping_Address Shipping_Method  \
0     0.00            NaN   False              NaN             NaN   
1     0.00            NaN   False              NaN             NaN   
2     0.00            NaN   False              NaN             NaN   
3  1299.66            NaN    True              NaN             NaN   
4    72.64            NaN    True              NaN             NaN   

                                         Product_URL  
0  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2  https://www.amazon.com/Samsung-Galaxy-Express-...  
3                                                NaN  
4                                                NaN

*Payment_Method*<br>
Accepted:<br>
Cards [Credit + Debit]: American Express, Visa, MasterCard, Discover<br>
Digital: PayPal<br>
Cash<br>
Consideration: let's say payment for in-person purchase and below 69.99 are Cash, online between 69.99 - 149.99 are most-likely PayPal, online between 149.99 - 249.99 are most-likely either Discover or AE, online between 249.99 - 449.99 are most-likely MasterCard and others are most-likely Visa

In [113]:
payment = ["AMEX", "Visa", "MasterCard", "Discover", "PayPal"]

In [114]:
def set_pay_method(row):
    if not row["Online"] and (row['Sales'] > 0.00 and row['Sales'] < 66.99):
        return "Cash"
    elif row['Sales'] >= 66.99 and row['Sales'] < 149.99:
        return (random.choices(["Discover", "PayPal"], weights=[.3, .7]))[0]
    elif row['Sales'] >= 149.99 and row['Sales'] < 349.99:
        return (random.choices(payment, weights=[.35, .1, .1, .35, .1]))[0]
    elif row['Sales'] >= 349.99 and row['Sales'] < 549.99:
        return (random.choices(payment, weights=[.225, .25, .25, .225, .05]))[0]
    elif row['Sales'] >= 549.99 and row['Sales'] < 999.99:
        return (random.choices(payment, weights=[.125, .34, .4, .125, .01]))[0]
    elif row['Sales'] > 999.99:
        return (random.choices(payment, weights=[.125, .4, .34, .125, .01]))[0]

In [115]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Purchase_ID        4026 non-null   object        
 1   Purchase_Datetime  4026 non-null   datetime64[ns]
 2   Product_ID         4026 non-null   object        
 3   Category           4026 non-null   object        
 4   Sub_Category       4026 non-null   object        
 5   Product_Name       4026 non-null   object        
 6   Unit_Price         4026 non-null   object        
 7   Quantity           4026 non-null   object        
 8   Sales              4026 non-null   float64       
 9   Payment_Method     0 non-null      object        
 10  Online             4026 non-null   bool          
 11  Shipping_Address   0 non-null      object        
 12  Shipping_Method    0 non-null      object        
 13  Product_URL        1400 non-null   object        
dtypes: bool(

In [116]:
df5["Payment_Method"] = df5.apply(lambda x: set_pay_method(x), axis=1)

In [117]:
df5.head()

Purchase_ID   Purchase_Datetime           Product_ID  \
0  116P20201113230027405 2020-11-13 23:00:27  BPC-TA-595E364B36B8   
1  412P20200714112438561 2020-07-14 11:24:38   CPA-A-AD2CDBF8FB6A   
2  284P20200318130450913 2020-03-18 13:04:50     E-P-7B472F9085C3   
3  210O20200305225756148 2020-03-05 22:57:56     E-P-06BB0F1A9C0B   
4  514O20200703152043999 2020-07-03 15:20:43     E-A-2252DB4688C6   

                    Category         Sub_Category  \
0     Beauty & Personal Care  Tools & Accessories   
1  Cell Phones & Accessories          Accessories   
2                Electronics               Phones   
3                Electronics               Phones   
4                Electronics          Accessories   

                                        Product_Name Unit_Price Quantity  \
0                Rubies Black Eyelashes and Adhesive       7.15        0   
1  Disney Mickey Mouse Ice Cream D-Lish Treats Ph...       8.25        0   
2                   Samsung Galaxy Express, Gray 8GB     389.28        0   
3                                     Apple iPhone 5     649.83        2   
4               SanDisk Cruzer 64 GB USB Flash Drive      36.32        2   

     Sales Payment_Method  Online Shipping_Address Shipping_Method  \
0     0.00           None   False              NaN             NaN   
1     0.00           None   False              NaN             NaN   
2     0.00           None   False              NaN             NaN   
3  1299.66           Visa    True              NaN             NaN   
4    72.64       Discover    True              NaN             NaN   

                                         Product_URL  
0  https://www.amazon.com/Rubies-Black-Eyelashes-...  
1  https://www.amazon.com/Disney-Mickey-Mouse-D-L...  
2  https://www.amazon.com/Samsung-Galaxy-Express-...  
3                                                NaN  
4                                                NaN

*Shipping_Address*<br>
Only where Online is True<br>
We gonna just go ahead and use `Faker` for this task.

In [118]:
fake = Faker(locale="en_US.UTF-8")
print(fake.address())

037 Williams Village
Lake Tony, OR 11726


In [119]:
df5.loc[df5["Shipping_Address"] == ""].head()

Empty DataFrame
Columns: [Purchase_ID, Purchase_Datetime, Product_ID, Category, Sub_Category, Product_Name, Unit_Price, Quantity, Sales, Payment_Method, Online, Shipping_Address, Shipping_Method, Product_URL]
Index: []

I also do want some addresses to show up multiple times. So, I'm going to save some addresses in a set and then distribute them over the records (online) in the DF. Say 1 dupes every 4 addresses.

In [120]:
# to calculate how many records
df5["Online"].value_counts()[1]

912

In [121]:
# to calculate how many different addresses to generate
np.uint8(np.ceil(df5["Online"].value_counts()[1] / 4))

228

In [122]:
def gen_addr():
    addrs = []
    count = np.uint8(np.ceil(df5["Online"].value_counts()[1] / 4))
    for _ in range(count):
        addr = " ".join(fake.address().split("\n"))
        while "Box" in addr or "FPO" in addr:
            addr = " ".join(fake.address().split("\n"))
        addrs.append(addr)
    return addrs
addresses = gen_addr()

In [123]:
def gen_addr(row):
    return random.choice(addresses) if row["Online"] else ""

In [124]:
df5["Shipping_Address"] = df5.apply(lambda x: gen_addr(x), axis=1)

In [125]:
df5.loc[df5["Online"]].head()

Purchase_ID   Purchase_Datetime        Product_ID     Category  \
3  210O20200305225756148 2020-03-05 22:57:56  E-P-06BB0F1A9C0B  Electronics   
4  514O20200703152043999 2020-07-03 15:20:43  E-A-2252DB4688C6  Electronics   
5  672O20200428055718472 2020-04-28 05:57:18  E-A-91E63EDED439  Electronics   
6  856O20200616225341793 2020-06-16 22:53:41  E-M-514AFA771511  Electronics   
8  532O20200926092642232 2020-09-26 09:26:42  E-P-EC919031AC36  Electronics   

  Sub_Category                                       Product_Name Unit_Price  \
3       Phones                                     Apple iPhone 5     649.83   
4  Accessories               SanDisk Cruzer 64 GB USB Flash Drive      36.32   
5  Accessories  Logitech Wireless Boombox Speaker - portable -...      106.4   
6     Machines                               Okidata B400 Printer      343.2   
8       Phones           Wilson Electronics DB Pro Signal Booster      286.4   

  Quantity    Sales Payment_Method  Online  \
3        2  1299.66           Visa    True   
4        2    72.64       Discover    True   
5        2   212.80       Discover    True   
6        1   343.20           AMEX    True   
8        3   859.20     MasterCard    True   

                                    Shipping_Address Shipping_Method  \
3         488 Ross Knolls New Cynthiashire, IN 57576             NaN   
4   0892 Thomas Track Suite 119 North Mark, GA 52991             NaN   
5  981 Linda Plains Apt. 504 East Brandontown, MA...             NaN   
6  52514 Kevin Springs Apt. 544 Adrianview, DC 17812             NaN   
8      682 Kathy Dale Apt. 888 Lozanoburgh, NY 35770             NaN   

  Product_URL  
3         NaN  
4         NaN  
5         NaN  
6         NaN  
8         NaN

*Shipping_Method*<br>
Available: Local Pickup, Standard Shipping, Overnight<br>
Overnight method from 449.99

In [126]:
ship_methods = {"Local Pickup", "Standard Shipping", "Overnight"}

In [127]:
def set_ship_method(row):
    if not row["Online"]:
        return ""
    elif row['Sales'] >= 66.99 and row['Sales'] < 449.99:
        return (random.choices(["Local Pickup", "Standard Shipping"], weights=[.3, .7]))[0]
    elif row["Sales"] >= 449.99:
        return (random.choices(list(ship_methods), weights=[.1, .6, .3]))[0]

In [128]:
df5["Shipping_Method"] = df5.apply(lambda x: set_ship_method(x), axis=1)

In [129]:
df5.loc[df5["Online"]].head()

Purchase_ID   Purchase_Datetime        Product_ID     Category  \
3  210O20200305225756148 2020-03-05 22:57:56  E-P-06BB0F1A9C0B  Electronics   
4  514O20200703152043999 2020-07-03 15:20:43  E-A-2252DB4688C6  Electronics   
5  672O20200428055718472 2020-04-28 05:57:18  E-A-91E63EDED439  Electronics   
6  856O20200616225341793 2020-06-16 22:53:41  E-M-514AFA771511  Electronics   
8  532O20200926092642232 2020-09-26 09:26:42  E-P-EC919031AC36  Electronics   

  Sub_Category                                       Product_Name Unit_Price  \
3       Phones                                     Apple iPhone 5     649.83   
4  Accessories               SanDisk Cruzer 64 GB USB Flash Drive      36.32   
5  Accessories  Logitech Wireless Boombox Speaker - portable -...      106.4   
6     Machines                               Okidata B400 Printer      343.2   
8       Phones           Wilson Electronics DB Pro Signal Booster      286.4   

  Quantity    Sales Payment_Method  Online  \
3        2  1299.66           Visa    True   
4        2    72.64       Discover    True   
5        2   212.80       Discover    True   
6        1   343.20           AMEX    True   
8        3   859.20     MasterCard    True   

                                    Shipping_Address    Shipping_Method  \
3         488 Ross Knolls New Cynthiashire, IN 57576          Overnight   
4   0892 Thomas Track Suite 119 North Mark, GA 52991       Local Pickup   
5  981 Linda Plains Apt. 504 East Brandontown, MA...  Standard Shipping   
6  52514 Kevin Springs Apt. 544 Adrianview, DC 17812  Standard Shipping   
8      682 Kathy Dale Apt. 888 Lozanoburgh, NY 35770       Local Pickup   

  Product_URL  
3         NaN  
4         NaN  
5         NaN  
6         NaN  
8         NaN

We'll have to add a Shipping_Fee feature, and therefore a Total one which will be the sum of Sales and Shipping_Fee.<br>
But for now let's just focus on quantity and sales.

*Quantity*<br>
Consideration: most expensive products are less likely to be purchased in multiples<br>
And given that we already have the prices we can just go and randomly set a quantity, say between 1 and 5, ranging from cheapest to most expensive products.

Let's first convert the Unit_Price column to float

In [130]:
df5["Unit_Price"] = df5["Unit_Price"].astype(np.float64)
# and round the values
df5["Unit_Price"] = df5["Unit_Price"].apply(lambda x: round(x, 2))

In [131]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4026 entries, 0 to 4025
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   Purchase_ID        4026 non-null   object        
 1   Purchase_Datetime  4026 non-null   datetime64[ns]
 2   Product_ID         4026 non-null   object        
 3   Category           4026 non-null   object        
 4   Sub_Category       4026 non-null   object        
 5   Product_Name       4026 non-null   object        
 6   Unit_Price         4026 non-null   float64       
 7   Quantity           4026 non-null   object        
 8   Sales              4026 non-null   float64       
 9   Payment_Method     2626 non-null   object        
 10  Online             4026 non-null   bool          
 11  Shipping_Address   4026 non-null   object        
 12  Shipping_Method    4026 non-null   object        
 13  Product_URL        1400 non-null   object        
dtypes: bool(

In [132]:
df5.loc[df5["Sales"] > 0.00].groupby(["Quantity"]).mean()["Unit_Price"]

Quantity
1     43.556693
2     58.942680
3     51.987424
4     76.967231
5     48.868366
6     39.427919
7     40.066173
8     44.844074
9     44.840313
10    18.353333
11    57.201111
12    54.543333
13    80.880000
14    37.841429
Name: Unit_Price, dtype: float64

So what this [above] basically shows us is the mean unit price for products by quantity

Let's create a new temp DF to keep track of those results

In [133]:
df_q = df5.loc[df5["Sales"] > 0.00].groupby(["Quantity"]).describe()["Unit_Price"]
df_q

count       mean         std   min      25%     50%       75%  \
Quantity                                                                  
1         257.0  43.556693   96.405231  0.44   5.3400  12.590   40.9900   
2         597.0  58.942680  155.696377  0.54   4.8900  12.000   48.5800   
3         590.0  51.987424  166.918509  0.42   4.8075  12.495   45.4800   
4         325.0  76.967231  296.784095  0.58   5.1800  13.580   48.7800   
5         361.0  48.868366  110.760264  0.34   4.7700  12.880   39.9900   
6         149.0  39.427919   97.832067  1.11   4.8100  10.390   25.9800   
7         162.0  40.066173   82.807975  0.59   5.0000  11.605   37.4775   
8          81.0  44.844074   71.011750  1.19   4.5800  11.960   55.6800   
9          64.0  44.840312   73.737722  0.36   4.3700  10.025   38.5775   
10         15.0  18.353333   28.868904  1.11   2.6750   8.390   15.4800   
11          9.0  57.201111   64.712281  6.23  13.9800  33.560   51.1800   
12          3.0  54.543333   58.831310  6.98  21.6500  36.320   78.3250   
13          6.0  80.880000  134.872282  1.04   3.4525   7.055  100.6800   
14          7.0  37.841429   46.478560  0.98   3.6900  15.970   59.4900   

              max  
Quantity           
1         1039.99  
2         2399.99  
3         3499.99  
4         3499.99  
5         1016.79  
6          717.59  
7          666.25  
8          300.65  
9          310.39  
10         109.95  
11         179.99  
12         120.33  
13         335.99  
14         121.58

In [134]:
df_q.std()

count     210.026267
mean       15.915205
std        67.833886
min         2.161868
25%         5.128511
50%         8.775126
75%        21.018347
max      1210.887010
dtype: float64

In [135]:
price_qty_tag = {idx: round(val, 2) for idx, val in enumerate(df_q["std"].tolist(), start=1)}
price_qty_tag

{1: 96.41,
 2: 155.7,
 3: 166.92,
 4: 296.78,
 5: 110.76,
 6: 97.83,
 7: 82.81,
 8: 71.01,
 9: 73.74,
 10: 28.87,
 11: 64.71,
 12: 58.83,
 13: 134.87,
 14: 46.48}

In [136]:
df5.loc[df5["Sales"] > 0.00].groupby(["Product_Name"]).count()["Quantity"].sort_values(ascending=False).reset_index()

Product_Name  Quantity
0                         Easy-staple paper        16
1                                   Staples        15
2                   Staples in misc. colors        12
3                           Staple envelope        11
4                   Storex Dura Pro Binders         8
...                                     ...       ...
1187        Quality Park Security Envelopes         1
1188  Decoflex Hanging Personal Folder File         1
1189      PureGear Roll-On Screen Protector         1
1190             Project Tote Personal File         1
1191     netTALK DUO VoIP Telephone Service         1

[1192 rows x 2 columns]